File: bridge-logistics-jms/src/main/java/com/jesperancinha/bridgelogistics/services/BridgeOpeningService.java
package com.jesperancinha.bridgelogistics.services;

import com.jesperancinha.bridgelogistics.data.BridgeDto;
import com.jesperancinha.bridgelogistics.data.BridgeOpeningConflictDto;
import com.jesperancinha.bridgelogistics.data.BridgeOpeningTimesDto;
import org.paukov.combinatorics3.Generator;
import org.paukov.combinatorics3.IGenerator;

import java.util.*;
import java.util.stream.Collectors;

/**
 * This service manages the bridge opening times service. It detects conflict and handles main bridge functions
 */
public class BridgeOpeningService {


    private Map<BridgeDto, List<BridgeOpeningTimesDto>> openingTimes;

    /**
     * Initialize your service with an opening times map per bridge name
     *
     * @param openingTimes {@link Map}
     */
    BridgeOpeningService(Map<BridgeDto, List<BridgeOpeningTimesDto>> openingTimes) {
        this.openingTimes = openingTimes;

    }

    /**
     * Initialize your service with an unsorted opening times list
     *
     * @param allOpeningTimes {@link List}
     */
    BridgeOpeningService(List<BridgeOpeningTimesDto> allOpeningTimes) {
        this.openingTimes = allOpeningTimes.stream().collect(
                Collectors.groupingBy(BridgeOpeningTimesDto::getBridge));
    }

    /**
     * Detects all conflictual time slots in current bridge times
     *
     * @return All conflicts map per bridge, per opening times {@link Map}
     */
    public Map<BridgeDto, Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto>> detectAllConflicts() {
        Map<BridgeDto, Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto>> allConflicts = new HashMap<>();
        openingTimes.forEach((bridgeName, bridgeOeningTimes) ->
                Generator.combination(bridgeOeningTimes)
                        .simple(2)
                        .stream()
                        .map(combination -> Generator.permutation(combination).simple())
                        .flatMap(IGenerator::stream)
                        .forEach(comb -> findConflictInATwoListBridgeOpeningTimes(allConflicts, comb)));

        return allConflicts;
    }

    private void findConflictInATwoListBridgeOpeningTimes(Map<BridgeDto, Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto>> allConflicts, List<BridgeOpeningTimesDto> comb) {
        BridgeOpeningTimesDto opening1 = comb.get(0);
        BridgeOpeningTimesDto opening2 = comb.get(1);
        if (hasConflicts(opening1, opening2)) {
            handleConflicts(allConflicts, opening1, opening2);
        }
    }

    private void handleConflicts(Map<BridgeDto, Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto>> allConflicts, BridgeOpeningTimesDto opening1, BridgeOpeningTimesDto opening2) {
        BridgeDto bridge = opening1.getBridge();
        Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto> currentBridgeConflicts = createOrGetBridgeOpeningDtoBridgeOpeningConflictMap(allConflicts, bridge);
        BridgeOpeningConflictDto bridgeOpeningConflictDto = createOrGetBridgeOpeningConflict(currentBridgeConflicts, opening1);
        bridgeOpeningConflictDto.getRelatedOpeningTimes().addAll(Arrays.asList(opening1, opening2));
        sanitize(bridgeOpeningConflictDto);
        allConflicts.put(bridge, currentBridgeConflicts);
    }

    private BridgeOpeningConflictDto createOrGetBridgeOpeningConflict(Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto> currentBridgeConflicts, BridgeOpeningTimesDto opening1) {
        BridgeOpeningConflictDto bridgeOpeningConflictDto = currentBridgeConflicts.get(opening1);
        if (Objects.isNull(bridgeOpeningConflictDto)) {
            bridgeOpeningConflictDto = new BridgeOpeningConflictDto();
            bridgeOpeningConflictDto.setRelatedOpeningTimes(new ArrayList<>());
            currentBridgeConflicts.put(opening1, bridgeOpeningConflictDto);
        }
        return bridgeOpeningConflictDto;
    }

    private Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto> createOrGetBridgeOpeningDtoBridgeOpeningConflictMap(Map<BridgeDto, Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto>> allConflicts, BridgeDto bridge) {
        Map<BridgeOpeningTimesDto, BridgeOpeningConflictDto> currentBridgeConflicts = allConflicts.get(bridge);
        if (Objects.isNull(currentBridgeConflicts)) {
            currentBridgeConflicts = new HashMap<>();
        }
        return currentBridgeConflicts;
    }

    /**
     * This method prevents conflictual opening times to be
     *
     * @param currentBridgeConflicts {@link BridgeOpeningConflictDto}
     */
    private void sanitize(BridgeOpeningConflictDto currentBridgeConflicts) {
        currentBridgeConflicts
                .setRelatedOpeningTimes(currentBridgeConflicts.getRelatedOpeningTimes()
                        .stream()
                        .sorted((ot1, ot2) -> ot1.getOpeningTime().isAfter(ot2.getOpeningTime()) ? 1 : 0)
                        .distinct()
                        .collect(Collectors.toList()));
    }

    private boolean hasConflicts(BridgeOpeningTimesDto opening1, BridgeOpeningTimesDto opening2) {
        return hasConflictsByOrder(opening1, opening2) || hasConflictsByOrder(opening2, opening1);
    }

    private boolean hasConflictsByOrder(BridgeOpeningTimesDto opening1, BridgeOpeningTimesDto opening2) {
        if (opening1.getBridge().equals(opening2.getBridge())) {
            if (opening1.getClosingTime().isAfter((opening2.getOpeningTime())) && opening1.getOpeningTime().isBefore(opening2.getOpeningTime())) {
                return true;
            } else
                return opening1.getOpeningTime().isBefore(opening2.getOpeningTime()) && opening1.getClosingTime().isAfter(opening2.getClosingTime());
        }
        return false;
    }

}

File: bl-central-server/bl-web-app/src/main/kotlin/org/jesperancinha/logistics/web/services/BridgeOpeningService.kt
package org.jesperancinha.logistics.web.services

import org.jesperancinha.logistics.jpa.dao.OpeningTimeRepository
import org.jesperancinha.logistics.web.dto.BridgeDto
import org.jesperancinha.logistics.web.dto.BridgeOpeningConflictDto
import org.jesperancinha.logistics.web.dto.BridgeOpeningTimeDto
import org.jesperancinha.logistics.web.utils.GeoCalculator
import org.paukov.combinatorics3.Generator
import org.paukov.combinatorics3.IGenerator
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.time.Instant
import java.util.*
import java.util.stream.Collectors

/**
 * This service manages the bridge opening times service. It detects conflict and handles main bridge functions
 */
@Service
class BridgeOpeningService(
    private val openingTimeRepository: OpeningTimeRepository,
    private val geoCalculator: GeoCalculator
) {
    fun isBridgeOpen(lat: BigDecimal, lon: BigDecimal): Boolean {
        val (westLatitude, eastLatitude, northLongitude, southLongitude) = geoCalculator.calculateSquareBoundary(
            lat,
            lon,
            BigDecimal.ONE
        )
        val bridgeByLatAndLonUnderRadius = openingTimeRepository.findBridgeBySquareBoundaryUnderRadius(
            westLatitude, eastLatitude, northLongitude, southLongitude,
            Instant.now()
                .toEpochMilli()
        )
        return bridgeByLatAndLonUnderRadius != null && bridgeByLatAndLonUnderRadius.size == 1
    }

    /**
     * Initialize your service with an opening times map per bridge name
     *
     * @param openingTimes [Map]
     * @return
     */
    fun getAllConflicts(openingTimes: Map<BridgeDto?, List<BridgeOpeningTimeDto>?>): Map<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>> {
        return detectAllConflicts(openingTimes)
    }

    /**
     * Initialize your service with an unsorted opening times list
     *
     * @param allOpeningTimes [List]
     * @return
     */
    fun getAllConflicts(allOpeningTimes: MutableList<BridgeOpeningTimeDto>): Map<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>> {
        val openingTimes = allOpeningTimes.stream()
            .collect(Collectors.groupingBy(BridgeOpeningTimeDto::bridge))
        return detectAllConflicts(openingTimes)
    }

    /**
     * Detects all conflictual time slots in current bridge times
     *
     * @param openingTimes
     * @return All conflicts map per bridge, per opening times [Map]
     */
    fun detectAllConflicts(openingTimes: Map<BridgeDto?, List<BridgeOpeningTimeDto>?>): Map<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>> {
        val allConflicts: MutableMap<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>> = HashMap()
        openingTimes.forEach { (bridgeName: BridgeDto?, bridgeOeningTimes: List<BridgeOpeningTimeDto>?) ->
            Generator.combination(bridgeOeningTimes)
                .simple(2)
                .stream()
                .map { combination: List<BridgeOpeningTimeDto>? ->
                    Generator.permutation(combination)
                        .simple()
                }
                .flatMap { obj: IGenerator<List<BridgeOpeningTimeDto>> -> obj.stream() }
                .forEach { comb: List<BridgeOpeningTimeDto> ->
                    findConflictInATwoListBridgeOpeningTimes(
                        allConflicts,
                        comb
                    )
                }
        }
        return allConflicts
    }

    private fun findConflictInATwoListBridgeOpeningTimes(
        allConflicts: MutableMap<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>>,
        comb: List<BridgeOpeningTimeDto>
    ) {
        val opening1 = comb[0]
        val opening2 = comb[1]
        if (hasConflicts(opening1, opening2)) {
            handleConflicts(allConflicts, opening1, opening2)
        }
    }

    private fun handleConflicts(
        allConflicts: MutableMap<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>>,
        opening1: BridgeOpeningTimeDto,
        opening2: BridgeOpeningTimeDto
    ) {
        val bridge = opening1.bridge
        val currentBridgeConflicts = createOrGetBridgeOpeningDtoBridgeOpeningConflictMap(allConflicts, bridge)
        var bridgeOpeningConflictDto = createOrGetBridgeOpeningConflict(currentBridgeConflicts, opening1)
        bridgeOpeningConflictDto!!.relatedOpeningTimes
            .addAll(listOf(opening1, opening2))
        if (currentBridgeConflicts[opening1] != null) {
            bridgeOpeningConflictDto = sanitize(bridgeOpeningConflictDto)
            currentBridgeConflicts[opening1] = bridgeOpeningConflictDto
        }
        allConflicts[bridge] = currentBridgeConflicts
    }

    private fun createOrGetBridgeOpeningConflict(
        currentBridgeConflicts: MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>,
        opening1: BridgeOpeningTimeDto
    ): BridgeOpeningConflictDto? {
        var bridgeOpeningConflictDto = currentBridgeConflicts[opening1]
        if (Objects.isNull(bridgeOpeningConflictDto)) {
            bridgeOpeningConflictDto = BridgeOpeningConflictDto(
                relatedOpeningTimes = mutableListOf()
            )
            currentBridgeConflicts[opening1] = bridgeOpeningConflictDto
        }
        return bridgeOpeningConflictDto
    }

    private fun createOrGetBridgeOpeningDtoBridgeOpeningConflictMap(
        allConflicts: Map<BridgeDto, MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?>>,
        bridge: BridgeDto
    ): MutableMap<BridgeOpeningTimeDto, BridgeOpeningConflictDto?> {
        return allConflicts[bridge] ?: HashMap()
    }

    /**
     * This method prevents conflictual opening times to be
     *
     * @param currentBridgeConflicts [BridgeOpeningConflictDto]
     * @return
     */
    private fun sanitize(currentBridgeConflicts: BridgeOpeningConflictDto?): BridgeOpeningConflictDto {
        return BridgeOpeningConflictDto(
            message = currentBridgeConflicts!!.message,
            relatedOpeningTimes =
            currentBridgeConflicts.relatedOpeningTimes
                .stream()
                .sorted { ot1: BridgeOpeningTimeDto, ot2: BridgeOpeningTimeDto ->
                    if (ot1.openingTime
                            .isAfter(ot2.openingTime)
                    ) 1 else 0
                }
                .distinct()
                .collect(Collectors.toList()))
    }

    private fun hasConflicts(opening1: BridgeOpeningTimeDto, opening2: BridgeOpeningTimeDto): Boolean {
        return hasConflictsByOrder(opening1, opening2) || hasConflictsByOrder(opening2, opening1)
    }

    private fun hasConflictsByOrder(opening1: BridgeOpeningTimeDto, opening2: BridgeOpeningTimeDto): Boolean {
        return if (opening1.bridge
            == opening2.bridge
        ) {
            if (opening1.closingTime
                    .isAfter(opening2.openingTime) && opening1.openingTime
                    .isBefore(opening2.openingTime)
            ) {
                true
            } else opening1.openingTime
                .isBefore(opening2.openingTime) && opening1.closingTime
                .isAfter(opening2.closingTime)
        } else false
    }
}